src/backend/utils/cache/plancache.c

```diff
@@ -941,6 +972,12 @@ BuildCachedPlan(CachedPlanSource *plansource, List *qlist,
 	 */
 	plist = pg_plan_queries(qlist, plansource->cursor_options, boundParams);
 
+	/* check if plangen switch appeared */
+	if (!plansource->plangen_switched)
```

Comment:
Maybe it is better to encapsulate all this logic into new API like `void detect_plangen_switch(plan, &plansource)`?

Reply:
Generally, I'm not into solutions, which gives full access to all struct members, if they don't need it. Unless such solutions need an access to much of them.
In my next commit I decided to even not use plan structure inside `find_plangen_switch`, but code around is full of opposite solutions. So, OK. Let it be.